File: kotlin/engine/src/main/kotlin/org/hpb/engine/nostr/Event.kt
package org.hpb.engine.nostr

import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.int
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.long
import org.hpb.engine.Secp
import org.hpb.engine.hex
import org.hpb.engine.hexBytes
import org.hpb.engine.sha256

/** A NIP-01 event; identity keys are the same secp256k1 x-only keys as descriptors. */
data class NostrEvent(
    val id: String,
    val pubkey: String,
    val createdAt: Long,
    val kind: Int,
    val tags: List<List<String>>,
    val content: String,
    val sig: String,
) {
    fun tagValue(name: String): String? =
        tags.firstOrNull { it.size >= 2 && it[0] == name }?.get(1)
}

object Events {
    private fun serializeForId(
        pubkey: String,
        createdAt: Long,
        kind: Int,
        tags: List<List<String>>,
        content: String,
    ): String = JsonArray(
        listOf(
            JsonPrimitive(0),
            JsonPrimitive(pubkey),
            JsonPrimitive(createdAt),
            JsonPrimitive(kind),
            JsonArray(tags.map { tag -> JsonArray(tag.map(::JsonPrimitive)) }),
            JsonPrimitive(content),
        ),
    ).toString()

    fun sign(
        privkey: ByteArray,
        kind: Int,
        tags: List<List<String>>,
        content: String,
        createdAt: Long = System.currentTimeMillis() / 1000,
    ): NostrEvent {
        val pubkey = Secp.xonlyHex(privkey)
        val id = sha256(serializeForId(pubkey, createdAt, kind, tags, content).toByteArray())
        return NostrEvent(
            id = id.hex(),
            pubkey = pubkey,
            createdAt = createdAt,
            kind = kind,
            tags = tags,
            content = content,
            sig = Secp.schnorrSign(id, privkey).hex(),
        )
    }

    fun verify(event: NostrEvent): Boolean {
        val id = sha256(
            serializeForId(event.pubkey, event.createdAt, event.kind, event.tags, event.content)
                .toByteArray(),
        )
        if (id.hex() != event.id) return false
        return runCatching {
            Secp.schnorrVerify(event.sig.hexBytes(), id, event.pubkey.hexBytes())
        }.getOrDefault(false)
    }

    fun toJson(event: NostrEvent): JsonObject = JsonObject(
        mapOf(
            "id" to JsonPrimitive(event.id),
            "pubkey" to JsonPrimitive(event.pubkey),
            "created_at" to JsonPrimitive(event.createdAt),
            "kind" to JsonPrimitive(event.kind),
            "tags" to JsonArray(event.tags.map { tag -> JsonArray(tag.map(::JsonPrimitive)) }),
            "content" to JsonPrimitive(event.content),
            "sig" to JsonPrimitive(event.sig),
        ),
    )

    fun fromJson(json: JsonObject): NostrEvent = NostrEvent(
        id = json.getValue("id").jsonPrimitive.content,
        pubkey = json.getValue("pubkey").jsonPrimitive.content,
        createdAt = json.getValue("created_at").jsonPrimitive.long,
        kind = json.getValue("kind").jsonPrimitive.int,
        tags = json.getValue("tags").jsonArray.map { tag ->
            tag.jsonArray.map { it.jsonPrimitive.content }
        },
        content = json.getValue("content").jsonPrimitive.content,
        sig = json.getValue("sig").jsonPrimitive.content,
    )
}

File: ios/HpbCore/Sources/HpbCore/Event.swift
import Foundation

/// A NIP-01 event; identity keys are the same secp256k1 x-only keys as the
/// escrow descriptors.
public struct NostrEvent: Equatable, Sendable {
    public let id: String
    public let pubkey: String
    public let createdAt: Int64
    public let kind: Int
    public let tags: [[String]]
    public let content: String
    public let sig: String

    public init(
        id: String, pubkey: String, createdAt: Int64, kind: Int,
        tags: [[String]], content: String, sig: String
    ) {
        self.id = id
        self.pubkey = pubkey
        self.createdAt = createdAt
        self.kind = kind
        self.tags = tags
        self.content = content
        self.sig = sig
    }

    public func tagValue(_ name: String) -> String? {
        tags.first { $0.count >= 2 && $0[0] == name }?[1]
    }
}

public enum Events {
    private static func serializeForId(
        pubkey: String, createdAt: Int64, kind: Int, tags: [[String]], content: String
    ) -> String {
        Cj.write(.arr([
            Cj.num(0),
            .str(pubkey),
            Cj.num(createdAt),
            Cj.num(kind),
            .arr(tags.map { tag in .arr(tag.map(J.str)) }),
            .str(content),
        ]))
    }

    public static func sign(
        _ privkey: [UInt8], kind: Int, tags: [[String]], content: String, createdAt: Int64
    ) throws -> NostrEvent {
        let pubkey = try Secp.xonlyHex(privkey)
        let id = sha256(serializeForId(
            pubkey: pubkey, createdAt: createdAt, kind: kind, tags: tags, content: content
        ))
        return NostrEvent(
            id: id.hex,
            pubkey: pubkey,
            createdAt: createdAt,
            kind: kind,
            tags: tags,
            content: content,
            sig: try Secp.schnorrSign(id, privkey).hex
        )
    }

    public static func verify(_ event: NostrEvent) -> Bool {
        let id = sha256(serializeForId(
            pubkey: event.pubkey, createdAt: event.createdAt, kind: event.kind,
            tags: event.tags, content: event.content
        ))
        guard id.hex == event.id else { return false }
        guard let sig = try? event.sig.hexBytes(),
              let xonly = try? event.pubkey.hexBytes() else { return false }
        return Secp.schnorrVerify(sig, id, xonly)
    }

    public static func toJson(_ event: NostrEvent) -> J {
        .obj([
            ("id", .str(event.id)),
            ("pubkey", .str(event.pubkey)),
            ("created_at", Cj.num(event.createdAt)),
            ("kind", Cj.num(event.kind)),
            ("tags", .arr(event.tags.map { tag in .arr(tag.map(J.str)) })),
            ("content", .str(event.content)),
            ("sig", .str(event.sig)),
        ])
    }

    public static func fromJson(_ json: J) throws -> NostrEvent {
        NostrEvent(
            id: try json.s("id"),
            pubkey: try json.s("pubkey"),
            createdAt: try json.l("created_at"),
            kind: try json.i("kind"),
            tags: try json.a("tags").map { tag in
                guard case .arr(let items) = tag else { throw HpbError.invalid("bad tag") }
                return try items.map {
                    guard let value = $0.stringValue else { throw HpbError.invalid("bad tag") }
                    return value
                }
            },
            content: try json.s("content"),
            sig: try json.s("sig")
        )
    }
}
